File: Java/java_lists/src/predo/javacore/classes/List.java
package predo.javacore.classes;

public class List {
    private int num;
    private List next;
    private static List head;
    private static List tail;

    public void push(int n){
        List aux, aux2; //Auxiliares para inserção

        if(List.head == null){ //Primeira isnerção
            this.num = n;
            this.next = null;
            changeHead(this); //Chama o método estático para realizar a alteração na head;
            //Se você passa apenas this como parametro, você passa o objeto atual inteiro como parametro!
            changeTail(this);
        }

        else if(List.head.next == null){
            if(n >= List.head.num){
                List.head.next = new List();
                aux = List.head.next; //Facilita no acesso do proximo elemento
                aux.num = n;
                aux.next = null;
                changeTail(aux);
            }

            else{
                aux = new List();
                aux.num = n;
                aux.next = List.head;
                changeTail(List.head);
                changeHead(aux);
            }
        }

        else if(n < List.head.num){
            aux = new List();
            aux.num = n;
            aux.next = List.head;
            changeHead(aux);
        }

        else if(n > List.tail.num){
            aux = new List();
            aux.num = n;
            aux.next = null;
            List.tail.next = aux;
            changeTail(aux);
        }

        else{
            aux = List.head;
            boolean complete = false;
            while(aux != null && !complete){
                if(n >= aux.num && n <= aux.next.num){
                    aux2 = new List();
                    aux2.num = n;
                    aux2.next = aux.next;
                    aux.next = aux2;
                    complete = true;
                }

                aux = aux.next;
            }
        }
    }

    public void pop(int n) {
        List aux, aux2;
        boolean removed = false;

        if(List.head == null) {
            System.out.println("Lista vazizia! Impossível remoção.");
        }

        else{
            if(n == List.head.num){ //Remoção do head
                aux = List.head;
                changeHead(List.head.next);
                aux = null;
                removed = true;
                System.gc(); //Comando que força a inicialização do garbage collector
            }

            else if(n == List.tail.num){ //Remoção do tail
                aux = List.head;
                while(aux.next != List.tail){
                    aux = aux.next;
                }
                aux2 = aux.next;
                changeTail(aux);
                aux.next = null;
                aux2 = null;
                removed = true;
                System.gc();
            }

            else{
                aux = List.head;
                while(aux.next != null && !removed){
                    if(aux.next.num == n){
                        aux2 = aux.next;
                        aux.next = aux.next.next;
                        aux2 = null;
                        removed = true;
                        System.gc();
                    }
                    aux = aux.next;
                }
            }

            if(removed) {
                System.out.println("Item removido com sucesso!");
            }
            else{
                System.out.println("Item não encontrado");
            }
        }
    }

    private static void changeHead(List newHead) { //Muda o valor estático de head
        List.head = newHead;
    }

    private static void changeTail(List newTail) { //Muda o valor estático de tail
        List.tail = newTail;
    }

    public void print(){
        List aux = List.head;
        if(head == null){
            System.out.println("Lista vazia! Nenhum elemento pode ser mostrado");
        }
        else {
            while(aux != null){
                System.out.println(aux.num);
                aux = aux.next;
            }
        }
    }
}

File: C/mias_listas/list.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

typedef struct nodes {
    char nome[64];
    struct nodes *next;
}node;

void pop(node **head, node **tail, char *name){
    node *aux, *aux2;
    bool removido = false;
    if((*head) == NULL)
        printf("\nNada pode ser removido\n");
    else{
        if((*head) == (*tail) && strcmp((*head)->nome, name) == 0){
            aux = (*head);
            (*head) = NULL;
            (*tail) = NULL;
            free(aux);
        }
        else if(strcmp((*head)->nome, name) == 0){
            aux = (*head);
            (*head) = (*head)->next;
            free(aux);
        }
        else{
            aux = (*head);
            while (aux != NULL && !removido) {
                if(strcmp(aux->next->nome, name) == 0){
                    aux2 = aux->next;
                    aux->next = aux->next->next;
                    free(aux2);
                    removido = true;
                }
                aux = aux->next;
            }

            if(removido)
                printf("Elemento removido com sucesso!\n");
            else
                printf("Elemento não encontrado!\n");
        }
    }
}

void push(node **head, node **tail, char *name){
    node *new = (node*)malloc(sizeof(node));
    node *aux = (*head);
    bool inserido = false;

    if((*head) == NULL){//Lista vazia
        printf("\n1\n");
        strcpy(new->nome, name);
        new->next = NULL;
        (*head) = new;
        (*tail) = new;
    }

    else if((*head) == (*tail)){ //Apenas um elemento
        if(strcmp(name, (*head)->nome) < 0){
            printf("\n2.1\n");
            strcpy(new->nome, name);
            new->next = (*head);
            (*head)->next = NULL;
            (*tail) = (*head);
            (*head) = new;
        }

        else if(strcmp(name, (*head)->nome) >= 0){
            printf("\n2.2\n");
            strcpy(new->nome, name);
            new->next = NULL;
            (*head)->next = new;
            (*tail) = new;
        }
    }

    else if(strcmp(name, (*head)->nome) < 0){
        strcpy(new->nome, name);
        new->next = (*head);
        (*head) = new;
    }

    else if(strcmp(name, (*tail)->nome) > 0){ //Maior que o tail
        printf("\n3\n");
        strcpy(new->nome, name);
        new->next = NULL;
        (*tail)->next = new;
        (*tail) = new;
    }

    else{ //Resto
        while (aux->next != NULL && !inserido){
            printf("\n4\n");
            if(strcmp(name, aux->nome) > 0 && strcmp(name, aux->next->nome) <= 0){
                strcpy(new->nome, name);
                new->next = aux->next;
                aux->next = new;
                inserido = true;
            }
            aux = aux->next;
        }
    }
}

void show(node **head){
    node *aux = (*head);
    int op;
    while (aux != NULL){
        printf("\n\nNome: %s\n", aux->nome);
        printf("Next: %p\n", aux->next);
        printf("(1)-Next || (2)-Sair: ");
        scanf("%d", &op);
        switch (op){
        case 1:
            aux = aux->next;
            break;
        
        case 2:
            aux = NULL;
            break;
        
        default:
            break;
        }
    }
    
}

int main(){
    node *head = NULL;
    node *tail = NULL;
    char nome[64];
    char input[64];

    for(int i = 0; i < 5; i++){
        setbuf(stdin, NULL);
        scanf("%[^\n]s", nome);
        push(&head, &tail, nome);
    }
    show(&head);
    setbuf(stdin, NULL);
    scanf("%[^\n]s", input);
    pop(&head, &tail, input);;
    show(&head);
    setbuf(stdin, NULL);
    scanf("%[^\n]s", input);
    pop(&head, &tail, input);
    show(&head);
    setbuf(stdin, NULL);
    scanf("%[^\n]s", input);
    pop(&head, &tail, input);
    show(&head);
    return 0;
}